File: tools/ebpf/custom_wt_ebpf/wt_memtrace/wt_memtrace.c
#ifdef asm_inline
#undef asm_inline
#define asm_inline asm
#endif

#include <uapi/linux/ptrace.h>

struct alloc_init_t {
    u64 size;
    void *retp;
};

struct alloc_info_t {
    u64 size;
    u64 timestamp_ns;
    int stack_id;
    u64 tgid_pid;
};

BPF_HASH(sizes, u64, struct alloc_init_t, 1000000);
BPF_HASH(allocs, u64, struct alloc_info_t, 1000000);
BPF_STACK_TRACE(stack_traces, 10240);

static inline int
gen_alloc_enter(struct pt_regs *ctx, size_t alloc_size, void *retp)
{
    // Sample every ~100 allocation
    u64 ts = bpf_ktime_get_ns();
    if (ts % 100 != 0)
        return 0;

    struct alloc_init_t init = {0};
    u64 pid = bpf_get_current_pid_tgid();
    init.size = alloc_size;
    init.retp = retp;
    sizes.update(&pid, &init);
    return 0;
}

static inline int
gen_alloc_exit2(struct pt_regs *ctx)
{
    u64 pid = bpf_get_current_pid_tgid();
    struct alloc_init_t *init = sizes.lookup(&pid);
    struct alloc_info_t info = {0};
    if (init == 0)
        return 0; // missed alloc entry
    void *address;
    bpf_probe_read(&address, sizeof(address), init->retp);
    u64 addr = (u64)address;
    info.size = init->size;
    sizes.delete(&pid);
    if (address != 0) {
        info.timestamp_ns = bpf_ktime_get_ns();
        info.stack_id = stack_traces.get_stackid(ctx, BPF_F_USER_STACK);
        info.tgid_pid = pid;
        allocs.update(&addr, &info);
    }
    return 0;
}

static inline int
gen_free_enter(struct pt_regs *ctx, void *address)
{
    u64 addr = (u64)address;
    struct alloc_info_t *info = allocs.lookup(&addr);
    if (info == 0)
        return 0;
    allocs.delete(&addr);
    return 0;
}

int
wt_malloc_enter(struct pt_regs *ctx, void *session, size_t bytes_to_allocate, void *retp)
{
    return gen_alloc_enter(ctx, bytes_to_allocate, retp);
}

int
wt_malloc_exit(struct pt_regs *ctx)
{
    return gen_alloc_exit2(ctx);
}

int
wt_free_enter(struct pt_regs *ctx)
{
    void *address;
    bpf_probe_read(&address, sizeof(address), (void *)PT_REGS_PARM2(ctx));
    return gen_free_enter(ctx, address);
}

int
wt_calloc_enter(struct pt_regs *ctx, void *session, size_t number, size_t size, void *retp)
{
    return gen_alloc_enter(ctx, number * size, retp);
}

int
wt_calloc_exit(struct pt_regs *ctx)
{
    return gen_alloc_exit2(ctx);
}

int
wt_realloc_enter(struct pt_regs *ctx, void *session, size_t *bytes_allocated_ret,
  size_t bytes_to_allocate, void *retp)
{
    void *address;
    bpf_probe_read(&address, sizeof(address), retp);
    gen_free_enter(ctx, address);
    return gen_alloc_enter(ctx, bytes_to_allocate, retp);
}

int
wt_realloc_exit(struct pt_regs *ctx)
{
    return gen_alloc_exit2(ctx);
}

File: tools/ebpf/custom_wt_ebpf/wt_memtrace/wt_memtrace.py
#!/usr/bin/python

from __future__ import print_function
from bcc import BPF
from datetime import datetime
import threading, os
import json, collections, time

# Minimum latency threshold required to report stack
mem_log_interval = 1 # 1 seconds
num_stacks = 5 # 1 seconds
min_age_ns = 1e6 * 500 # Prune allocations younger than 5ms

class Allocation(object):
    def __init__(self, stack, size):
        self.stack = stack
        self.count = 1
        self.size = size

    def update(self, size):
        self.count += 1
        self.size += size

class MemTrace():
    def __init__(self, wt_lib, functions):
        self.functions = functions
        self.traces = collections.OrderedDict()
        self.traces_count = {}

        # Initialise BPF program
        src = os.path.dirname(__file__) + '/wt_memtrace.c'
        self.b = BPF(src_file=src)

        self.b.attach_uprobe(name=wt_lib, sym='__wt_malloc', fn_name='wt_malloc_enter')
        self.b.attach_uretprobe(name=wt_lib, sym='__wt_malloc', fn_name='wt_malloc_exit')
        self.b.attach_uprobe(name=wt_lib, sym='__wt_calloc', fn_name='wt_calloc_enter')
        self.b.attach_uretprobe(name=wt_lib, sym='__wt_calloc', fn_name='wt_calloc_exit')
        self.b.attach_uprobe(name=wt_lib, sym='__wt_realloc', fn_name='wt_realloc_enter')
        self.b.attach_uretprobe(name=wt_lib, sym='__wt_realloc', fn_name='wt_realloc_exit')
        self.b.attach_uprobe(name=wt_lib, sym='__wt_realloc_noclear', fn_name='wt_realloc_enter')
        self.b.attach_uretprobe(name=wt_lib, sym='__wt_realloc_noclear', fn_name='wt_realloc_exit')
        self.b.attach_uprobe(name=wt_lib, sym='__wt_free_int', fn_name='wt_free_enter')

        # Create our memtrace file
        script_dir = os.path.dirname(__file__)
        trace_filename = script_dir + '/../../stats/memstrace.stat'
        os.makedirs(os.path.dirname(trace_filename), exist_ok=True)
        self.trace_out = open(trace_filename,'w')

    def log_outstanding_allocations(self, trace_log):
        alloc_info = {}
        allocs = self.b["allocs"]
        stack_traces = self.b["stack_traces"]
        for address, info in allocs.items():
            if BPF.monotonic_time() - min_age_ns < info.timestamp_ns:
                continue
            if info.stack_id < 0:
                continue
            if info.stack_id in alloc_info:
                alloc_info[info.stack_id].update(info.size)
            else:
                stack = list(stack_traces.walk(info.stack_id))
                combined = []
                symbols_found = False
                for addr in stack:
                    sym = self.b.sym(addr, info.tgid_pid, show_offset=True)
                    sym = sym.decode('utf8')
                    if ("[unknown]" not in sym):
                        symbols_found = True
                    combined.append(self.b.sym(addr, info.tgid_pid, show_offset=True))
                if symbols_found:
                    alloc_info[info.stack_id] = Allocation(combined, info.size)

        if(len(alloc_info.values()) >= 1):
            log_output = ""
            log_output += "[%s] Top %d stacks with outstanding allocations:\n" % (datetime.now().strftime("%H:%M:%S"), min(len(alloc_info.values()), num_stacks))
            to_show = sorted(alloc_info.values(),key=lambda a: a.size)[-num_stacks:]
            for alloc in to_show:
                log_output += "\t%d bytes in %d allocations from stack\n\t\t%s\n" % (alloc.size, alloc.count, b"\n\t\t".join(alloc.stack).decode("ascii"))
            trace_log.send(log_output.encode())
            self.trace_out.write(log_output)

    def enter_trace(self, exit_event, trace_log):
        log_time = time.time()
        while not exit_event.is_set():
            self.log_outstanding_allocations(trace_log)
            time.sleep(mem_log_interval)

def customTraceThread(functions, wt_lib, exit_event, sock):
    memTracer = MemTrace(wt_lib, functions)
    memTracer.enter_trace(exit_event, sock)
